src/main/java/software/amazon/event/kafkaconnector/offloading/S3EventBridgeEventDetailValueOffloading.java

```diff
@@ -0,0 +1,181 @@
+    } catch (Exception e) {
+      throw new IllegalArgumentException(format("Invalid JSON Path '%s'.", value));
+    }
+
+    if (!jsonPath.getPath().startsWith("$['detail']['value']")) {
+      throw new IllegalArgumentException(
+          format("JSON Path must start with '%s' but is '%s'.", JSON_PATH_PREFIX, value));
+    }
+    if (!jsonPath.isDefinite()) {
+      throw new IllegalArgumentException(
+          format("JSON Path must be definite but '%s' is not.", value));
+    }
+    // rewrite $.detail -> $,
+    // because PutEventsRequestEntry#detail is the sub document of $.detail
+    return JsonPath.compile("$" + jsonPath.getPath().substring("$['detail']".length()));
+  }
+
+  public static void validateJsonPath(String jsonPathExp) {
+    transformToPutEventsRequestEntry(jsonPathExp);
+  }
+
+  @Override
+  public EventBridgeEventDetailValueOffloadingResult apply(
+      final List<MappedSinkRecord<PutEventsRequestEntry>> putEventsRequestEntries) {
+
+    var result =
+        putEventsRequestEntries.stream()
+            .map(this::apply)
+            .collect(partitioningBy(EventBridgeResult::isSuccess));
+
+    var success = result.get(true).stream().map(EventBridgeResult::success).collect(toList());
+    var errors = result.get(false).stream().map(EventBridgeResult::failure).collect(toList());
+
+    return new EventBridgeEventDetailValueOffloadingResult(success, errors);
+  }
+
+  private EventBridgeResult<PutEventsRequestEntry> apply(
+      final MappedSinkRecord<PutEventsRequestEntry> item) {
+
+    var sinkRecord = item.getSinkRecord();
+    var putEventsRequestEntry = item.getValue();
+
+    try {
+      var documentContext = JsonPath.parse(putEventsRequestEntry.detail(), configuration);
+
+      var value = documentContext.read(jsonPathRemove);
+      if (value != null) {
+        var payload =
+            value instanceof Map || value instanceof List
+                ? JsonPath.parse(value).jsonString()
+                : value.toString();
+        documentContext.delete(jsonPathRemove).put(jsonPathAdd, "dataref", putS3(payload));
+      }
+      documentContext.put(jsonPathAdd, "datarefPath", jsonPathExp);
+
+      return success(
+          sinkRecord, putEventsRequestEntry.copy(it -> it.detail(documentContext.jsonString())));
+
+    } catch (final ThrowingFunctionApplyException e) {
+      var unwrapped = e.getCause();
+      if (unwrapped instanceof ExecutionException
+          || unwrapped instanceof InterruptedException
+          || unwrapped instanceof TimeoutException) {
+
+        var cause = unwrapped.getCause();
+        if (cause instanceof S3Exception && ((S3Exception) cause).statusCode() < 500) {
+          return failure(sinkRecord, panic(e));
+        }
+        return failure(sinkRecord, retry(e));
+      }
+      return failure(sinkRecord, panic(unwrapped));
+    }
+  }
+
+  private String putS3(final String payload) {
+    logger.debug("Found JSON at '{}'.", jsonPathRemove.getPath());
```

Comment:
nit: IMHO we mostly use TRACE level across the codebase, not debug?

Reply:
Will check.